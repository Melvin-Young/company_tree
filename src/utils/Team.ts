export interface IStaff {
  id: number,
  name: string,
  team: string
}

export interface ITeam {
  name: string,
  getParent: () => ITeam,
  setParent: (parent: ITeam) => void,
  getChildren: () => Map<string, ITeam>,
  getChild: (child: string) => ITeam,
  addChild: (child: ITeam) => Map<string, ITeam>,
  hasChild: (child: string) => boolean,
  getStaff: () => Map<number, IStaff>,
  addStaff: (staff: IStaff) => Map<number, IStaff>,
  getNestedStaff: () => Map<number, IStaff>,
  addNestedStaff: (staff: IStaff) => Map<number, IStaff>
}

export default class Team implements ITeam {
  public name: string;
  private parent: ITeam;
  private children: Map<string, ITeam>;
  private staff: Map<number, IStaff>;
  private nestedStaff: Map<number, IStaff>;

  constructor(name: string) {
    if (!name) {
      throw new Error('Name is required');
    }
    this.name = name;
    this.parent = {} as ITeam;
    this.children = new Map();
    this.staff = new Map();
    this.nestedStaff = new Map();
  }

  public getParent(): ITeam {
    return this.parent;
  }

  public setParent(parent: ITeam): void {
    this.parent = parent;
  }

  public getChildren(): Map<string, ITeam> {
    return this.children;
  }

  public getChild(child: string): ITeam {
    return this.children.get(child) as ITeam;
  }

  public addChild(child: ITeam): Map<string, ITeam> {
    return this.children.set(child.name, child);
  }

  public hasChild(child: string): boolean {
    return this.children.has(child);
  }

  public getStaff(): Map<number, IStaff> {
    return this.staff;
  }

  public addStaff(staff: IStaff): Map<number, IStaff> {
    return this.staff.set(staff.id, staff);
  }

  public getNestedStaff(): Map<number, IStaff> {
    return this.nestedStaff;
  }

  public addNestedStaff(staff: IStaff): Map<number, IStaff> {
    return this.nestedStaff.set(staff.id, staff);
  }
}