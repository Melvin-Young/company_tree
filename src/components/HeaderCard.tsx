/* tslint:disable no-unused-expression no-console jsx-no-lambda */

import React, { Component } from 'react';
import ConfigurableButton from '../buttons/ConfigurableButton';
import { ITeam } from '../utils/Team';
import { IStaff } from '../utils/Team';

interface IProps {
  team: ITeam,
  displayNewTeam: (team: ITeam) => void,
  showStaff: () => void,
  showTeams: () => void
}
interface IState {
  team: ITeam
}
class HeaderCard extends Component<IProps, IState> {
  public constructor(props: IProps) {
    super(props);
    this.state = {
      team: props.team
    }
  }

  public render() {
    const { displayNewTeam, showStaff, showTeams, team } = this.props;
    const staff = team.getStaff();
    const staffContainer: any[] = [];
    staff.size && staff.forEach((staffMember: IStaff) => {
      staffContainer.push(<li key={staffMember.id}> {staffMember.name} </li>)
    });

    return (
      <div className='card header-card'>
        <div onClick={ () => displayNewTeam(team.getParent()) }>
          <img src={require('./../images/brainstorming.jpg')} alt="Avatar" className="card-picture" />
          <h2> {team.name} Team</h2>
          <div className="team-member-list">
              <ul> {staffContainer} </ul>
          </div>
        </div>
          <div className="button-container">
            <ConfigurableButton buttonText="Show Staff" onClickFunc={() => showStaff()}/>
            <ConfigurableButton buttonText="Show Teams" onClickFunc={() => showTeams()}/>
          </div>
      </div>
    )
  }
}

export default HeaderCard;