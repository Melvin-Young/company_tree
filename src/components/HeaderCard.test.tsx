/* tslint:disable no-unused-expression jsx-no-lambda */

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';

import Team from '../utils/Team';
import HeaderCard from './HeaderCard';

Enzyme.configure({ adapter: new Adapter()});

const dummyFunc = jest.fn();
const dummyTeam = new Team('Dummy Team');

describe('TeamCard', () => {
  it('should render', () => {
    const component = shallow(<HeaderCard team={dummyTeam} setHeader={() => undefined} triggerShowStaff={dummyFunc} triggerShowTeams={dummyFunc} />)
    expect(component).toMatchSnapshot();
  });
});