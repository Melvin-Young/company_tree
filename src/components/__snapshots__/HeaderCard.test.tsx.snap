// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamCard should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HeaderCard
    setHeader={[Function]}
    team={
        Team {
            "children": Map {},
            "name": "Dummy Team",
            "nestedStaff": Map {},
            "parent": Object {},
            "staff": Map {},
          }
    }
    triggerShowStaff={[Function]}
    triggerShowTeams={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          onClick={[Function]}
>
          <img
                    alt="Avatar"
                    className="card-picture"
                    src="brainstorming.jpg"
          />
          <h2>
                     
                    Dummy Team
                     Team
          </h2>
          <div
                    className="team-member-list"
          >
                    <ul>
                               
                               
                    </ul>
          </div>
</div>,
        <div
          className="button-container"
>
          <ConfigurableButton
                    buttonText="Show Staff"
                    onClickFunc={[Function]}
          />
          <ConfigurableButton
                    buttonText="Show Teams"
                    onClickFunc={[Function]}
          />
</div>,
      ],
      "className": "card header-card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="Avatar"
              className="card-picture"
              src="brainstorming.jpg"
/>,
            <h2>
               
              Dummy Team
               Team
</h2>,
            <div
              className="team-member-list"
>
              <ul>
                             
                             
              </ul>
</div>,
          ],
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "Avatar",
              "className": "card-picture",
              "src": "brainstorming.jpg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                "Dummy Team",
                " Team",
              ],
            },
            "ref": null,
            "rendered": Array [
              " ",
              "Dummy Team",
              " Team",
            ],
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <ul>
                 
                 
</ul>,
              "className": "team-member-list",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  Array [],
                  " ",
                ],
              },
              "ref": null,
              "rendered": Array [
                " ",
                " ",
              ],
              "type": "ul",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ConfigurableButton
              buttonText="Show Staff"
              onClickFunc={[Function]}
/>,
            <ConfigurableButton
              buttonText="Show Teams"
              onClickFunc={[Function]}
/>,
          ],
          "className": "button-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "buttonText": "Show Staff",
              "onClickFunc": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "buttonText": "Show Teams",
              "onClickFunc": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            onClick={[Function]}
>
            <img
                        alt="Avatar"
                        className="card-picture"
                        src="brainstorming.jpg"
            />
            <h2>
                         
                        Dummy Team
                         Team
            </h2>
            <div
                        className="team-member-list"
            >
                        <ul>
                                     
                                     
                        </ul>
            </div>
</div>,
          <div
            className="button-container"
>
            <ConfigurableButton
                        buttonText="Show Staff"
                        onClickFunc={[Function]}
            />
            <ConfigurableButton
                        buttonText="Show Teams"
                        onClickFunc={[Function]}
            />
</div>,
        ],
        "className": "card header-card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="Avatar"
                className="card-picture"
                src="brainstorming.jpg"
/>,
              <h2>
                 
                Dummy Team
                 Team
</h2>,
              <div
                className="team-member-list"
>
                <ul>
                                 
                                 
                </ul>
</div>,
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "Avatar",
                "className": "card-picture",
                "src": "brainstorming.jpg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  "Dummy Team",
                  " Team",
                ],
              },
              "ref": null,
              "rendered": Array [
                " ",
                "Dummy Team",
                " Team",
              ],
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <ul>
                   
                   
</ul>,
                "className": "team-member-list",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    " ",
                    Array [],
                    " ",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  " ",
                  " ",
                ],
                "type": "ul",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ConfigurableButton
                buttonText="Show Staff"
                onClickFunc={[Function]}
/>,
              <ConfigurableButton
                buttonText="Show Teams"
                onClickFunc={[Function]}
/>,
            ],
            "className": "button-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "buttonText": "Show Staff",
                "onClickFunc": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "buttonText": "Show Teams",
                "onClickFunc": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
